spring-pulsar/src/main/java/org/springframework/pulsar/listener/reactive/DefaultReactivePulsarMessageListenerContainer.java

```diff
@@ -0,0 +1,215 @@
+				doStart();
+			}
+		}
+	}
+
+	@Override
+	public void stop() {
+		synchronized (this.lifecycleMonitor) {
+			if (isRunning()) {
+				doStop();
+			}
+		}
+	}
+
+	public ReactiveMessageConsumerBuilderCustomizer<T> getConsumerCustomizer() {
+		return this.consumerCustomizer;
+	}
+
+	@Override
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	public void setConsumerCustomizer(ReactiveMessageConsumerBuilderCustomizer consumerCustomizer) {
+		this.consumerCustomizer = consumerCustomizer;
+	}
+
+	private void doStart() {
+		this.pipeline = startPipeline(this.pulsarContainerProperties);
+		setRunning(true);
+	}
+
+	public void doStop() {
+		try {
+			this.logger.info("Closing this reactive pipeline.");
+			this.pipeline.close();
+		}
+		catch (Exception e) {
+			this.logger.error(e, () -> "Error closing Pulsar Reactive pipeline.");
+		}
+	}
+
+	@Override
+	public void destroy() {
+
+	}
+
+	@SuppressWarnings({ "unchecked" })
+	private ReactiveMessagePipeline startPipeline(ReactivePulsarContainerProperties containerProperties) {
+		ReactiveMessageConsumerBuilderCustomizer<T> customizer = (builder) -> {
+			if (containerProperties.getSubscriptionType() != null) {
+				builder.subscriptionType(containerProperties.getSubscriptionType());
+			}
+			if (containerProperties.getSubscriptionName() != null) {
+				builder.subscriptionName(containerProperties.getSubscriptionName());
+			}
+			if (containerProperties.getTopics() != null) {
+				List<String> topicNames = Arrays.asList(containerProperties.getTopics());
+				if (!topicNames.isEmpty()) {
+					builder.topicNames(topicNames);
+				}
+			}
+			if (containerProperties.getTopicsPattern() != null) {
+				builder.topicsPattern(Pattern.compile(containerProperties.getTopicsPattern()));
+			}
+		};
+
+		List<ReactiveMessageConsumerBuilderCustomizer<T>> customizers = new ArrayList<>();
+		customizers.add(customizer);
+		if (this.consumerCustomizer != null) {
+			customizers.add(this.consumerCustomizer);
+		}
+
+		ReactiveMessageConsumer<T> consumer = getReactivePulsarConsumerFactory()
+				.createConsumer((Schema<T>) containerProperties.getSchema(), customizers);
+		ReactiveMessagePipelineBuilder<T> pipelineBuilder = ApiImplementationFactory
+				.createReactiveMessageHandlerPipelineBuilder(consumer);
+		Object messageHandler = containerProperties.getMessageHandler();
+		ReactiveMessagePipeline pipeline;
+		if (messageHandler instanceof ReactivePulsarStreamingHandler<?>) {
+			pipeline = pipelineBuilder
+					.streamingMessageHandler(((ReactivePulsarStreamingHandler<T>) messageHandler)::received).build();
+		}
+		else {
+			ReactiveMessagePipelineBuilder.OneByOneMessagePipelineBuilder<T> messagePipelineBuilder = pipelineBuilder
+					.messageHandler(((ReactivePulsarMessageHandler<T>) messageHandler)::received)
+					.handlingTimeout(Duration.ofMillis(containerProperties.getHandlingTimeoutMillis()));
+			if (containerProperties.getConcurrency() > 0) {
```

Comment:
Concurrency is a little different than concurrency in the imperative PMLC.
In PMLC, concurrency means that multiple identical containers are used that have each one consumer with one messageListener.
Here, concurrency is in the `reactor` meaning where only one consumer is used that will dispatch messages to multiple concurrent messageHandler.